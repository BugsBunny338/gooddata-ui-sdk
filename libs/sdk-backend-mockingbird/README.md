# Mocking bird

This package contains test support and mock implementations of various concepts in the sdk-backend-spi.
