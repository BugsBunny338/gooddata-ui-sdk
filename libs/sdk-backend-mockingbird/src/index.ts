// (C) 2019-2020 GoodData Corporation

import {
    dummyBackend,
    dummyBackendEmptyData,
    dummyDataView,
    dummyDataFacade,
} from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView, dummyDataFacade };

export { recordedBackend } from "./recordedBackend";

export {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
} from "./recordedBackend/types";

export {
    recordedDataView,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
} from "./recordedBackend/execution";

export {
    LegacyExecutionRecording,
    legacyRecordedBackend,
    legacyRecordedDataFacade,
} from "./legacyRecordedBackend";
