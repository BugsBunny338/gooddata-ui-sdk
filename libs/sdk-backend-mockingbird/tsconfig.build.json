{
    "extends": "./tsconfig.json",
    "compilerOptions": {
        "baseUrl": null,
        "paths": null
    },
    "include": ["src/index.ts"]
}
