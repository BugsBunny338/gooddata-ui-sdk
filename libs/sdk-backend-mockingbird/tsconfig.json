{
    "compilerOptions": {
        "esModuleInterop": true,
        "target": "es5",
        "module": "commonjs",
        "lib": ["es6", "dom"],
        "declaration": true,
        "outDir": "./dist",
        "strict": true,
        "noImplicitAny": true,
        "strictNullChecks": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "baseUrl": "..",
        "paths": {
            "@gooddata/sdk-model": ["sdk-model/src"],
            "@gooddata/sdk-backend-base": ["sdk-backend-base/src"],
            "@gooddata/sdk-backend-spi": ["sdk-backend-spi/src"]
        }
    },
    "exclude": ["node_modules", "dist"]
}
