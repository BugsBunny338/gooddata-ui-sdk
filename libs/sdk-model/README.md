# SDK domain models

## License

Copyright (C) 2007-2020, GoodData(R) Corporation. All rights reserved.

For more information, please see [LICENSE](https://github.com/gooddata/gooddata-ui-sdk/blob/master/libs/sdk-model/LICENSE)
