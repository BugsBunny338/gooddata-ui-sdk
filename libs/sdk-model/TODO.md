Functionality:

-   Add builders for the types in this package
-   Move & add type guards
-   Move general purpose functions working with these types into this package

Tests:

-   Move existing tests, add new tests

Documentation:

-   Migration guide: text filters are no longer defined using textFilter indicator
-   100% doc coverage needed here
