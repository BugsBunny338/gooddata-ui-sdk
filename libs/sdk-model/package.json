{
    "name": "@gooddata/sdk-model",
    "version": "8.0.0-alpha.0",
    "author": "GoodData",
    "description": "GoodData SDK TypeScript skeleton",
    "repository": "https://github.com/gooddata/gooddata-ui-sdk/tree/master/libs/sdk-model",
    "license": "BSD-3-Clause",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "src",
        "LICENSE"
    ],
    "config": {
        "tslint": "./{src,test}/**/*.ts"
    },
    "scripts": {
        "dev": "tsc --watch",
        "build": "rm -rf dist && tsc -p tsconfig.build.json && npm run api-extractor",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --config jest.ci.json",
        "tslint": "tslint -p . -t verbose $npm_package_config_tslint",
        "tslint-ci": "mkdir -p ./ci/results && tslint -p . -t checkstyle -o ./ci/results/tslint-results.xml $npm_package_config_tslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "validate": "tsc --noEmit && npm run tslint && npm run prettier-check",
        "validate-ci": "tsc --noEmit && npm run tslint-ci && npm run prettier-check"
    },
    "jest": {
        "preset": "ts-jest",
        "setupFilesAfterEnv": [
            "<rootDir>/jest.setup.ts"
        ],
        "testRegex": "\\.test.ts$",
        "testPathIgnorePatterns": [
            "/node_modules/",
            "/dist/"
        ],
        "collectCoverageFrom": [
            "src/**/*.ts"
        ],
        "coveragePathIgnorePatterns": [
            "/node_modules/"
        ],
        "collectCoverage": true,
        "coverageReporters": [
            "html",
            "json"
        ],
        "globals": {
            "ts-jest": {
                "packageJson": "package.json"
            }
        }
    },
    "dependencies": {
        "lodash": "^4.17.15",
        "stringify-object": "^2.4.0",
        "spark-md5": "^3.0.0",
        "ts-invariant": "^0.4.4"
    },
    "devDependencies": {
        "@gooddata/tslint-config": "1.0.0",
        "@microsoft/api-extractor": "^7.3.8",
        "@types/jest": "^24.0.6",
        "@types/lodash": "^4.14.123",
        "@types/node": "^10.14.18",
        "@types/spark-md5": "^3.0.1",
        "@types/stringify-object": "^3.2.0",
        "jest": "^24.7.1",
        "jest-junit": "^3.0.0",
        "prettier": "~1.18.2",
        "ts-jest": "^24.1.0",
        "tslint": "^5.14.0",
        "tslint-config-prettier": "^1.18.0",
        "typescript": "3.6.3"
    }
}
