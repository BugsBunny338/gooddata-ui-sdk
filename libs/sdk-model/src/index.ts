// (C) 2019 GoodData Corporation

export {
    IAttribute,
    isAttribute,
    attributeId,
    AttributePredicate,
    anyAttribute,
    idMatchAttribute,
    attributeFingerprint,
    attributesFind,
    attributeUri,
    attributeIdentifier,
} from "./attribute";

export {
    Identifier,
    IObjUriQualifier,
    IObjIdentifierQualifier,
    IObjLocalIdentifierQualifier,
    ObjQualifier,
    ObjQualifierWithLocal,
    isUriQualifier,
    isIdentifierQualifier,
} from "./base";

export {
    IDimension,
    isDimension,
    dimensionTotals,
    newTwoDimensional,
    dimensionFingerprint,
    MeasureGroupIdentifier,
} from "./base/dimension";

export { TotalType, ITotal, isTotal, totalIsNative } from "./base/totals";

export {
    SortDirection,
    IAttributeSortItem,
    SortItem,
    IMeasureSortItem,
    LocatorItem,
    IAttributeLocatorItem,
    IMeasureLocatorItem,
    isMeasureLocator,
    isAttributeLocator,
    isMeasureSort,
    isAttributeSort,
    newMeasureSort,
    newAttributeSort,
    sortFingerprint,
} from "./base/sort";

export {
    IAttributeElementsByRef,
    IAttributeElementsByValue,
    AttributeElements,
    IPositiveAttributeFilter,
    INegativeAttributeFilter,
    IAbsoluteDateFilter,
    IRelativeDateFilter,
    IFilter,
    IDateFilter,
    IAttributeFilter,
    isAbsoluteDateFilter,
    isRelativeDateFilter,
    isPositiveAttributeFilter,
    isNegativeAttributeFilter,
    isDateFilter,
    isAttributeFilter,
    isAttributeElementsByRef,
    isAttributeElementsByValue,
    attributeElementsIsEmpty,
    filterIsEmpty,
    filterFingerprint,
} from "./filter";

export {
    IMeasureDefinitionType,
    IMeasureDefinition,
    ArithmeticMeasureOperator,
    IArithmeticMeasureDefinition,
    IPoPMeasureDefinition,
    IMeasure,
    MeasureAggregation,
    IPreviousPeriodMeasureDefinition,
    IPreviousPeriodDateDataSet,
    isMeasure,
    isSimpleMeasure,
    isPoPMeasure,
    isPreviousPeriodMeasure,
    isArithmeticMeasure,
    isMeasureDefinition,
    isPoPMeasureDefinition,
    isPreviousPeriodMeasureDefinition,
    isArithmeticMeasureDefinition,
    measureId,
    MeasurePredicate,
    anyMeasure,
    idMatchMeasure,
    measureDoesComputeRatio,
    measureFingerprint,
    measureUri,
    measureIdentifier,
} from "./measure";

export {
    AttributeOrMeasure,
    IBucket,
    isBucket,
    idMatchBucket,
    anyBucket,
    MeasureInBucket,
    AttributeInBucket,
    newBucket,
    bucketIsEmpty,
    bucketAttributes,
    bucketAttribute,
    bucketMeasures,
    bucketTotals,
    bucketItems,
    bucketsAttributes,
    bucketsMeasures,
    bucketsById,
    bucketsIsEmpty,
    bucketsItems,
    bucketsFind,
    bucketsFindAttribute,
    bucketsFindMeasure,
    bucketsTotals,
    ComputeRatioRule,
    computeRatioRules,
    BucketPredicate,
} from "./buckets";

export {
    GuidType,
    RGBType,
    IColor,
    IColorItem,
    IColorMappingProperty,
    IColorPalette,
    IColorPaletteItem,
    IGuidColorItem,
    IRGBColorItem,
} from "./colors";

export {
    IInsight,
    IVisualizationClass,
    VisualizationProperties,
    isInsight,
    insightMeasures,
    insightHasMeasures,
    insightAttributes,
    insightHasAttributes,
    insightHasDataDefined,
    insightProperties,
    insightBuckets,
    insightSorts,
    insightBucket,
    insightTotals,
    insightFilters,
} from "./insight";
