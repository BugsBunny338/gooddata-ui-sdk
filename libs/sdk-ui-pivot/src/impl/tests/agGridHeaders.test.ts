// (C) 2007-2020 GoodData Corporation

import { ReferenceRecordings, ReferenceLdm } from "@gooddata/reference-workspace";
import { DataViewFirstPage, recordedDataView } from "@gooddata/sdk-backend-mockingbird";
import {
    DataViewFacade,
    IAttributeDescriptor,
    IResultHeader,
    isAttributeDescriptor,
    isMeasureDescriptor,
} from "@gooddata/sdk-backend-spi";
import { attributeLocalId, IAttributeSortItem, IMeasureSortItem } from "@gooddata/sdk-model";
import {
    assortDimensionDescriptors,
    getAttributeSortItemFieldAndDirection,
    getColumnHeaders,
    getFields,
    getMeasureSortItemFieldAndDirection,
    getMinimalRowData,
    getRowHeaders,
    headerToGrid,
    identifyHeader,
    identifyResponseHeader,
    mergeHeaderEndIndex,
    shouldMergeHeaders,
} from "../agGridHeaders";

const fixture = recordedDataView(
    ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
    DataViewFirstPage,
);

describe("identifyHeader", () => {
    it("should return correct field key for an attribute header", () => {
        expect(identifyHeader(fixture.allHeaders()[0][0][0])).toMatchSnapshot();
    });

    it("should return correct field key for a measure header", () => {
        expect(identifyHeader(fixture.allHeaders()[1][1][0])).toMatchSnapshot();
    });
});
describe("identifyResponseHeader", () => {
    it("should return correct field key for an attribute response header", () => {
        expect(identifyResponseHeader(fixture.dimensionItemDescriptors(0)[0])).toMatchSnapshot();
    });
});

describe("headerToGrid", () => {
    it("should return correct grid header for an attribute header with correct prefix", () => {
        expect(headerToGrid(fixture.allHeaders()[0][0][0], "prefix_")).toMatchSnapshot();
    });

    it("should return correct grid header for a measure header with correct prefix", () => {
        expect(headerToGrid(fixture.allHeaders()[1][1][0], "prefix_")).toMatchSnapshot();
    });
});

describe("getColumnHeaders", () => {
    it("should return hierarchical column headers", () => {
        expect(
            getColumnHeaders(fixture.allHeaders()[1], fixture.dimensionItemDescriptors(1)),
        ).toMatchSnapshot();
    });
});

describe("getRowHeaders", () => {
    it("should return an array of grid headers", () => {
        expect(
            getRowHeaders(fixture.dimensionItemDescriptors(0) as IAttributeDescriptor[], {}, false),
        ).toMatchSnapshot();
    });
    it("should return an array of grid headers with row group settings and extended by custom options", () => {
        expect(
            getRowHeaders(
                fixture.dimensionItemDescriptors(0) as IAttributeDescriptor[],
                { type: "custom" },
                true,
            ),
        ).toMatchSnapshot();
    });
});

describe("getFields", () => {
    const Fixtures: Array<[string, DataViewFacade]> = [
        [
            "single measure, single column",
            recordedDataView(
                ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAndColumnAttributes,
                DataViewFirstPage,
            ),
        ],
        [
            "single measure, multiple columns",
            recordedDataView(
                ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithTwoRowAndTwoColumnAttributes,
                DataViewFirstPage,
            ),
        ],
        [
            "two measures, single columns",
            recordedDataView(
                ReferenceRecordings.Scenarios.PivotTable.TwoMeasuresWithRowAndColumnAttributes,
                DataViewFirstPage,
            ),
        ],
    ];

    it.each(Fixtures)("should correctly obtain column fields for %s", (_desc, facade) => {
        expect(getFields(facade.allHeaders()[1])).toMatchSnapshot();
    });
});

describe("assortDimensionDescriptors", () => {
    it("should return attribute and measure dimension headers", () => {
        const dimensions = fixture.dimensions();
        const { attributeDescriptors, measureDescriptors } = assortDimensionDescriptors(dimensions);
        expect(attributeDescriptors).toHaveLength(2);
        expect(attributeDescriptors.filter(header => isAttributeDescriptor(header))).toHaveLength(2);
        expect(measureDescriptors).toHaveLength(1);
        expect(measureDescriptors.filter(header => isMeasureDescriptor(header))).toHaveLength(1);
    });
});

describe("getMinimalRowData", () => {
    const NoMeasureData = recordedDataView(
        ReferenceRecordings.Scenarios.PivotTable.SingleAttribute,
        DataViewFirstPage,
    );
    const WithMeasureData = recordedDataView(
        ReferenceRecordings.Scenarios.PivotTable.SingleMeasureWithRowAttribute,
        DataViewFirstPage,
    );

    it("should return a two-dimensional array of empty values when no measure data are available", () => {
        const result = getMinimalRowData(NoMeasureData);
        const expectedLength = NoMeasureData.allHeaders()[0][0].length;

        expect(result.length).toEqual(expectedLength);
        expect(result.filter(e => e[0] === null).length).toEqual(expectedLength);
    });

    it("should return a identical data if measure data is available", () => {
        const result = getMinimalRowData(WithMeasureData);
        const expectedLength = WithMeasureData.twoDimData().length;

        expect(result.length).toEqual(expectedLength);
        expect(result).toEqual(WithMeasureData.twoDimData());
    });
});

describe("conversion from header matrix to hierarchy", () => {
    const alabamaHeader = {
        attributeHeaderItem: {
            uri: "/gdc/md/xms7ga4tf3g3nzucd8380o2bev8oeknp/obj/2210/elements?id=1",
            name: "Alabama",
        },
    };

    const californiaHeader = {
        attributeHeaderItem: {
            uri: "/gdc/md/xms7ga4tf3g3nzucd8380o2bev8oeknp/obj/2210/elements?id=2",
            name: "California",
        },
    };

    const year2017Header = {
        attributeHeaderItem: {
            uri: "/gdc/md/xms7ga4tf3g3nzucd8380o2bev8oeknp/obj/2212/elements?id=3",
            name: "2017",
        },
    };

    const costsHeader = {
        measureHeaderItem: {
            name: "Costs",
            order: 1,
        },
    };
    const revenuesHeader = {
        measureHeaderItem: {
            name: "Revenues",
            order: 2,
        },
    };

    const resultHeaderDimension: IResultHeader[][] = [
        [alabamaHeader, alabamaHeader, californiaHeader, californiaHeader],
        [year2017Header, year2017Header, year2017Header, year2017Header],
        [costsHeader, revenuesHeader, costsHeader, revenuesHeader],
    ];

    const stateHeaderIndex = 0;
    const yearHeaderIndex = 1;
    const measureGroupHeaderIndex = 2;

    describe("shouldMergeHeaders", () => {
        it("should return true for headers that are identical and have no ancestors", () => {
            expect(shouldMergeHeaders(resultHeaderDimension, stateHeaderIndex, 0)).toBe(true);
        });
        it("should return true for headers that are identical and have identical ancestors", () => {
            expect(shouldMergeHeaders(resultHeaderDimension, yearHeaderIndex, 0)).toBe(true);
        });
        it("should return false for headers that are identical but have at least one non-identical ancestor", () => {
            expect(shouldMergeHeaders(resultHeaderDimension, yearHeaderIndex, 1)).toBe(false);
        });
        it("should return false for headers that are not identical", () => {
            expect(shouldMergeHeaders(resultHeaderDimension, measureGroupHeaderIndex, 0)).toBe(false);
        });
    });

    describe("mergeHeaderCount", () => {
        it("should return correct index of last header in row that can be merged with the current one", () => {
            expect(mergeHeaderEndIndex(resultHeaderDimension, stateHeaderIndex, 0)).toBe(1);
            expect(mergeHeaderEndIndex(resultHeaderDimension, stateHeaderIndex, 1)).toBe(1);
            expect(mergeHeaderEndIndex(resultHeaderDimension, stateHeaderIndex, 2)).toBe(3);
            expect(mergeHeaderEndIndex(resultHeaderDimension, stateHeaderIndex, 3)).toBe(3);
        });
    });
});

describe("getAttributeSortItemFieldAndDirection", () => {
    const dimensions = fixture.dimensions();
    const { attributeDescriptors } = assortDimensionDescriptors(dimensions);
    const attributeSortItem: IAttributeSortItem = {
        attributeSortItem: {
            direction: "asc",
            attributeIdentifier: attributeLocalId(ReferenceLdm.Product.Name),
        },
    };
    it("should return matching key and direction from attributeDescriptors", () => {
        expect(getAttributeSortItemFieldAndDirection(attributeSortItem, attributeDescriptors)).toEqual([
            "a_1055",
            "asc",
        ]);
    });
});

describe("getMeasureSortItemFieldAndDirection", () => {
    const dimensions = fixture.dimensions();
    const { measureDescriptors } = assortDimensionDescriptors(dimensions);
    const measureSortItem: IMeasureSortItem = {
        measureSortItem: {
            direction: "desc",
            locators: [
                {
                    attributeLocatorItem: {
                        attributeIdentifier: "date.aam81lMifn6q",
                        element: "/gdc/md/xms7ga4tf3g3nzucd8380o2bev8oeknp/obj/2009/elements?id=1",
                    },
                },
                {
                    attributeLocatorItem: {
                        attributeIdentifier: "date.abm81lMifn6q",
                        element: "/gdc/md/xms7ga4tf3g3nzucd8380o2bev8oeknp/obj/2071/elements?id=1",
                    },
                },
                {
                    measureLocatorItem: {
                        measureIdentifier: "aaEGaXAEgB7U",
                    },
                },
            ],
        },
    };
    it("should return matching key and direction from measure descriptors", () => {
        expect(getMeasureSortItemFieldAndDirection(measureSortItem, measureDescriptors)).toEqual([
            "a_2009_1-a_2071_1-m_-1",
            "desc",
        ]);
    });
});
