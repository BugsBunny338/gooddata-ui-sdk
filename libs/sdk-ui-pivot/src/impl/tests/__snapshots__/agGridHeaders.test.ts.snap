// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getColumnHeaders should return hierarchical column headers 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "drillItems": Array [
          Object {
            "attributeHeaderItem": Object {
              "name": "East Coast",
              "uri": "/gdc/md/referenceworkspace/obj/1086/elements?id=460488",
            },
          },
          Object {
            "attributeHeader": Object {
              "formOf": Object {
                "identifier": "attr.owner.region",
                "name": "Region",
                "ref": Object {
                  "uri": "/gdc/md/referenceworkspace/obj/1086",
                },
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "identifier": "label.owner.region",
              "localIdentifier": "a_label.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1087",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
          },
          Object {
            "measureHeaderItem": Object {
              "format": "$#,##0.00",
              "identifier": "aangOxLSeztu",
              "localIdentifier": "m_aangOxLSeztu",
              "name": "Amount",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1271",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1271",
            },
          },
        ],
        "field": "a_1086_460488-m_0",
        "headerName": "Amount",
        "measureIndex": 0,
        "type": "MEASURE_COLUMN",
      },
    ],
    "drillItems": Array [],
    "field": "a_1086_460488",
    "headerName": "East Coast",
    "type": "COLUMN_ATTRIBUTE_COLUMN",
  },
  Object {
    "children": Array [
      Object {
        "drillItems": Array [
          Object {
            "attributeHeaderItem": Object {
              "name": "West Coast",
              "uri": "/gdc/md/referenceworkspace/obj/1086/elements?id=460489",
            },
          },
          Object {
            "attributeHeader": Object {
              "formOf": Object {
                "identifier": "attr.owner.region",
                "name": "Region",
                "ref": Object {
                  "uri": "/gdc/md/referenceworkspace/obj/1086",
                },
                "uri": "/gdc/md/referenceworkspace/obj/1086",
              },
              "identifier": "label.owner.region",
              "localIdentifier": "a_label.owner.region",
              "name": "Region",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1087",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1087",
            },
          },
          Object {
            "measureHeaderItem": Object {
              "format": "$#,##0.00",
              "identifier": "aangOxLSeztu",
              "localIdentifier": "m_aangOxLSeztu",
              "name": "Amount",
              "ref": Object {
                "uri": "/gdc/md/referenceworkspace/obj/1271",
              },
              "uri": "/gdc/md/referenceworkspace/obj/1271",
            },
          },
        ],
        "field": "a_1086_460489-m_0",
        "headerName": "Amount",
        "measureIndex": 0,
        "type": "MEASURE_COLUMN",
      },
    ],
    "drillItems": Array [],
    "field": "a_1086_460489",
    "headerName": "West Coast",
    "type": "COLUMN_ATTRIBUTE_COLUMN",
  },
]
`;

exports[`getFields should correctly obtain column fields for single measure, multiple columns 1`] = `
Array [
  "a_1091_166498-a_1086_460488-m_0",
  "a_1091_166498-a_1086_460489-m_0",
  "a_1091_166647-a_1086_460488-m_0",
  "a_1091_166647-a_1086_460489-m_0",
  "a_1091_168751-a_1086_460488-m_0",
  "a_1091_168751-a_1086_460489-m_0",
  "a_1091_166425-a_1086_460488-m_0",
  "a_1091_166425-a_1086_460489-m_0",
  "a_1091_166442-a_1086_460488-m_0",
  "a_1091_166442-a_1086_460489-m_0",
  "a_1091_166450-a_1086_460488-m_0",
  "a_1091_166450-a_1086_460489-m_0",
  "a_1091_165679-a_1086_460488-m_0",
  "a_1091_165679-a_1086_460489-m_0",
  "a_1091_165872-a_1086_460488-m_0",
  "a_1091_165872-a_1086_460489-m_0",
]
`;

exports[`getFields should correctly obtain column fields for single measure, single column 1`] = `
Array [
  "a_1086_460488-m_0",
  "a_1086_460489-m_0",
]
`;

exports[`getFields should correctly obtain column fields for two measures, single columns 1`] = `
Array [
  "a_1086_460488-m_0",
  "a_1086_460488-m_1",
  "a_1086_460489-m_0",
  "a_1086_460489-m_1",
]
`;

exports[`getRowHeaders should return an array of grid headers 1`] = `
Array [
  Object {
    "drillItems": Array [
      Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.product.id",
            "name": "Product",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1054",
          },
          "identifier": "label.product.id.name",
          "localIdentifier": "a_label.product.id.name",
          "name": "Product Name",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1055",
        },
      },
    ],
    "field": "a_1055",
    "headerName": "Product",
    "type": "ROW_ATTRIBUTE_COLUMN",
  },
]
`;

exports[`getRowHeaders should return an array of grid headers with row group settings and extended by custom options 1`] = `
Array [
  Object {
    "drillItems": Array [
      Object {
        "attributeHeader": Object {
          "formOf": Object {
            "identifier": "attr.product.id",
            "name": "Product",
            "ref": Object {
              "uri": "/gdc/md/referenceworkspace/obj/1054",
            },
            "uri": "/gdc/md/referenceworkspace/obj/1054",
          },
          "identifier": "label.product.id.name",
          "localIdentifier": "a_label.product.id.name",
          "name": "Product Name",
          "ref": Object {
            "uri": "/gdc/md/referenceworkspace/obj/1055",
          },
          "uri": "/gdc/md/referenceworkspace/obj/1055",
        },
      },
    ],
    "field": "a_1055",
    "headerName": "Product",
    "hide": true,
    "rowGroup": true,
    "type": "custom",
  },
]
`;

exports[`headerToGrid should return correct grid header for a measure header with correct prefix 1`] = `
Object {
  "field": "prefix_m_0",
  "headerName": "Amount",
}
`;

exports[`headerToGrid should return correct grid header for an attribute header with correct prefix 1`] = `
Object {
  "field": "prefix_a_1054_165678",
  "headerName": "CompuSci",
}
`;

exports[`identifyHeader should return correct field key for a measure header 1`] = `"m_0"`;

exports[`identifyHeader should return correct field key for an attribute header 1`] = `"a_1054_165678"`;

exports[`identifyResponseHeader should return correct field key for an attribute response header 1`] = `"a_1055"`;
